cpp/src/arrow/compute/exec/spilling_join.h

```diff
@@ -0,0 +1,129 @@
+  using ApplyBloomFilterCallback = std::function<Status(size_t, ExecBatch*)>;
+  using OutputBatchCallback = std::function<void(int64_t, ExecBatch)>;
+  using FinishedCallback = std::function<Status(int64_t)>;
+  using StartSpillingCallback = std::function<Status(size_t)>;
+  using PauseProbeSideCallback = std::function<void(int)>;
+  using ResumeProbeSideCallback = std::function<void(int)>;
+
+  struct CallbackRecord {
+    RegisterTaskGroupCallback register_task_group;
+    StartTaskGroupCallback start_task_group;
+    AddProbeSideHashColumn add_probe_side_hashes;
+    BloomFilterFinishedCallback bloom_filter_finished;
+    ApplyBloomFilterCallback apply_bloom_filter;
+    OutputBatchCallback output_batch;
+    FinishedCallback finished;
+    StartSpillingCallback start_spilling;
+    PauseProbeSideCallback pause_probe_side;
+    ResumeProbeSideCallback resume_probe_side;
+  };
+
+  Status Init(QueryContext* ctx, JoinType join_type, size_t num_threads,
+              SchemaProjectionMaps<HashJoinProjection>* proj_map_left,
+              SchemaProjectionMaps<HashJoinProjection>* proj_map_right,
+              std::vector<JoinKeyCmp>* key_cmp, Expression* filter,
+              PartitionedBloomFilter* bloom_filter, CallbackRecord callback_record,
+              bool is_swiss);
+
+  Status CheckSpilling(size_t thread_index, ExecBatch& batch);
+
+  Status OnBuildSideBatch(size_t thread_index, ExecBatch batch);
+  Status OnBuildSideFinished(size_t thread_index);
+
+  Status OnProbeSideBatch(size_t thread_index, ExecBatch batch);
+  Status OnProbeSideFinished(size_t thread_index);
+
+  Status OnBloomFiltersReceived(size_t thread_index);
+
+ private:
+  Status AdvanceSpillCursor(size_t thread_index);
+
+  // Builds the entire bloom filter for all 64 partitions.
+  Status BuildPartitionedBloomFilter(size_t thread_index);
+  Status PushBloomFilterBatch(size_t thread_index, int64_t batch_id);
+  // Builds a bloom filter for a single partition.
+  Status BuildNextBloomFilter(size_t thread_index);
+  Status OnBloomFilterFinished(size_t thread_index);
+  Status OnPartitionedBloomFilterFinished(size_t thread_index);
+
+  Status StartCollocatedJoins(size_t thread_index);
+  Status BeginNextCollocatedJoin(size_t thread_index);
+  Status BuildHashTable(size_t thread_index);
+  Status OnHashTableFinished(size_t thread_index);
+  Status OnProbeSideBatchReadBack(size_t thread_index, size_t batch_idx, ExecBatch batch);
+  Status OnProbingFinished(size_t thread_index);
+  Status OnCollocatedJoinFinished(int64_t num_batches);
+
+  QueryContext* ctx_;
+  size_t num_threads_;
+  CallbackRecord callbacks_;
+  bool is_swiss_;
+  PartitionedBloomFilter* bloom_filter_;
+  std::unique_ptr<BloomFilterBuilder> builder_;
+
+  // Backpressure toggling happens at most twice during execution. A value of 0 means
+  // we haven't toggled backpressure at all, value of 1 means we've paused, and value
+  // 2 means we've resumed.
+  std::atomic<int32_t> backpressure_counter_{0};
```

Comment:
This would probably need a followup. It might not be terribly difficult, we could probably use the StopToken functionality to pause the tasks associated with the spill file in the IO executor.

Reply:
I think a stop token is one-shot and we might need to pause / resume / pause / resume / etc (imagine a very slow consumer).  However, I agree that it can left for a follow-up.